drivers/serial/uart_nrfx_uarte.c

```diff
@@ -1551,27 +1586,28 @@ static void txstopped_isr(const struct device *dev)
 	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
 	unsigned int key;
 
-	if (config->flags & UARTE_CFG_FLAG_LOW_POWER) {
-		nrf_uarte_int_disable(uarte, NRF_UARTE_INT_TXSTOPPED_MASK);
-		key = irq_lock();
-		uarte_disable_locked(dev, UARTE_FLAG_LOW_POWER_TX, UARTE_FLAG_LOW_POWER_RX);
-		irq_unlock(key);
+	key = irq_lock();
 
-		if (!data->async->tx.len) {
-			return;
+	size_t amount = (data->async->tx.amount >= 0) ?
+			data->async->tx.amount : nrf_uarte_tx_amount_get(uarte);
+
+	if (IS_ENABLED(CONFIG_PM_DEVICE_RUNTIME)) {
+		nrf_uarte_int_disable(uarte, NRF_UARTE_INT_TXSTOPPED_MASK);
+		if (data->flags & UARTE_FLAG_POLL_OUT) {
+			pm_device_runtime_put(dev);
+			data->flags &= ~UARTE_FLAG_POLL_OUT;
 		}
+	} else if (LOW_POWER_ENABLED(config)) {
+		nrf_uarte_int_disable(uarte, NRF_UARTE_INT_TXSTOPPED_MASK);
+		uarte_disable_locked(dev, UARTE_FLAG_LOW_POWER_TX, UARTE_FLAG_LOW_POWER_RX);
 	}
```

Comment:
All of this is only for `uarte_nrfx_poll_out()`, right? Why not use the `UARTE_FLAG_POLL_OUT` flag to cover all this stuff here (and the related part of poll_out) and perform low-power disabling for a normal async transfer together with that `pm_device_runtime_put()` at the end of this function.
And similarly with `uarte_enable_locked()`, instead of calling it from `tx_start()`, why not call it together with `pm_device_runtime_get()` - separately in `start_tx_locked()` and `uarte_nrfx_poll_out()`? I think this would make things a bit clearer. And it will save some unnecessary low-power enabling and disabling when a transfer needs to be done in parts, through the TX cache buffer. There's one complication that I can see in such approach - if a TX transfer is pending because of an earlier poll_out, disabling would need to be skipped or re-enabling would need to be done before that pending transfer is started. This should not be that bad.

Reply:
I did not apply it yet. Started to work on that but first draft shown slight code size increase. I'm also afraid about missing some corner cases (current implementation was passing all tests). Though, i agree that it would be slightly better to avoid disabling/re-enabling peripheral when cache is used. I will see tomorrow if i manage to add it in this PR.

Reply:
This can be improved in a follow-up PR.